packages/engine/paima-rest/src/controllers/AchievementsController.ts

```diff
@@ -46,25 +57,54 @@ export class AchievementsController extends Controller {
     /** Comma-separated list. */
     @Query() name?: string
   ): Promise<PlayerAchievements> {
-    const player = await service().getPlayer(wallet);
-    const achievements = await service().getPlayerAchievements(wallet);
-    const nameSet = name ? new Set(name.split(',')) : null;
+    const names = name ? name.split(',') : [];
+    const player: Player = { wallet };
+
+    const db = EngineService.INSTANCE.getSM().getReadonlyDbConn();
+    const rows = await getAchievementProgress.run({ wallet, names }, db);
+
+    this.setHeader('Content-Language', 'en');
     return {
-      ...(await this.defaultValidity()),
-      ...(await service().getValidity()),
+      ...(await this.validity()),
       ...player,
-      completed: achievements.reduce((n, ach) => n + (ach.completed ? 1 : 0), 0),
-      achievements: nameSet ? achievements.filter(ach => nameSet.has(ach.name)) : achievements,
+      completed: rows.reduce((n, row) => n + (row.completed_date ? 1 : 0), 0),
+      achievements: rows.map(row => ({
+        name: row.name,
+        completed: Boolean(row.completed_date),
+        completedDate: row.completed_date ?? undefined,
+        completedRate: row.total
+          ? {
+              progress: row.progress ?? 0,
+              total: row.total,
+            }
+          : undefined,
+      })),
     };
   }
 
-  @Get('nft/{nft_address}')
+  @Get('erc/{erc}/{cde}/{token_id}')
   public async nft(
-    @Path() nft_address: string,
-    /** Comma-separated list. */
+    @Path() erc: string,
+    @Path() cde: string,
+    @Path() token_id: string,
     @Query() name?: string
   ): Promise<PlayerAchievements> {
-    const wallet = await service().getNftOwner(nft_address);
-    return await this.wallet(wallet, name);
+    const db = EngineService.INSTANCE.getSM().getReadonlyDbConn();
+    this.setHeader('Content-Language', 'en');
+
+    switch (erc) {
+      case 'erc721':
+        const wallet = await getNftOwner(db, cde, BigInt(token_id));
+        if (wallet) {
+          return await this.wallet(wallet, name);
+        }
+        break;
+      case 'erc6551':
+        // TODO
+        break;
+    }
+
+    this.setStatus(404);
```

Comment:
It might be better to provide a more helpful message based on the cde

i.e.
1. If it exists but it's the wrong type (not 721), say the type if not supported
2. If it doesn't exist, mention there is no primitive with that name

Reply:
Seems reasonable. Will worm the CDE list in here somehow